proposals/NNNN-objc-implementation.md

```diff
@@ -0,0 +1,323 @@
+
+@property (strong) UIViewController *frontViewController;
+@property (strong) UIViewController *backViewController;
+@property (assign,getter=isShowingFront) BOOL showingFront;
+
+- (instancetype)initWithFrontViewController:(UIViewController *)front backViewController:(UIVIewController *)back;
+
+@end
+
+@interface MYFlippableViewController (Animation)
+
+- (void)setShowingFront:(BOOL)isShowingFront animated:(BOOL)animated NS_SWIFT_NAME(setIsShowingFront(_:animated:));
+- (void)setFrontViewController:(UIViewController *)front animated:(BOOL)animated;
+- (void)setBackViewController:(UIViewController *)back animated:(BOOL)animated;
+
+@end
+
+@interface MYFlippableViewController (Actions)
+
+- (IBAction)flip:(id)sender;
+
+@end
+
+NS_HEADER_AUDIT_END(nullability, sendability)
+```
+
+And you would arrange for Swift to import it through an umbrella or bridging header. You would then write an `extension` for each `@interface` you wish to implement in Swift. For example, you could implement the main `@interface` (plus any visible class extensions) in Swift by writing:
+
+```swift
+@objc @implementation extension MYFlippableViewController {
+    ...
+}
+```
+
+And the `Animation` category by writing:
+
+```swift
+@objc(Animation) @implementation extension MYFlippableViewController {
+    ...
+}
+```
+
+Note that there is nothing special in the header which indicates that a given `@interface` is implemented in Swift. The header can use all of the usual Swift annotations—like `NS_SWIFT_NAME`, `NS_NOESCAPE` etc.—but they simply affect how the member is imported. Swift does not even require you to implement every declared `@interface` in Swift, so you can implement some parts of a class in Objective-C and others in Swift. But if you choose to implement a particular `@interface` in Swift, each Objective-C member in that `@interface` must be matched by a Swift declaration in the extension that has the same Swift name; these special members are called "member implementations".
+
+An `@objc @implementation extension` can contain four kinds of members:
+
+1. **Open, public, package, or internal `@objc` members** must be member implementations. Swift will give you an error if they don't match a member from the imported headers, so it will diagnose typos and mismatched Swift names.
+2. **Fileprivate or private `@objc` members** are helper methods (think `@IBAction`s or callback selectors). They must *not* match a member from the imported headers, but they are accessible from Objective-C by performing a selector or declaring them in a place that is not visible to Swift.
+3. **Members with a `final` modifier (or `@nonobjc` on an initializer)** are Swift-only and can use Swift-only types or features. These may be Swift-only implementation details (if `internal` or `private`) or Swift-only APIs (if `public` or `package`).
+4. **Members with an `override` modifier** override superclass members and function normally.
```

Comment:
I wonder if it's worth a brief couple sentences justifying these rules. It's not immediately obvious why each one exists (IMO (1) and (3) are especially non-obvious, I'm not confident I could articulate the exact motivation).

Reply:
The way I came up with to handle this is to treat category 1 (member implementations) separately from the other three and give an explicit justification for allowing each of the other three categories.

```suggestion
Specifically, member implementations must be non-`final`, not be overrides, and have an `open`, `public`, `package`, or `internal` access level. Every member implementation must match a member declared in the Objective-C header, and every member declared in the Objective-C header must have a matching member implementation. This ensures that everything declared by the header is correctly implemented without any accidental misspellings or type signature mismatches.

In addition to member implementations, an `@objc @implementation` extension can also contain three other kinds of members:

1. **Fileprivate or private non-`final` members** are helper methods (think `@IBAction`s or callback selectors). They must *not* match a member from the imported headers, but they are accessible from Objective-C by performing a selector or declaring them in a place that is not visible to Swift. (Objective-C `@implementation` blocks often declare private members like this, so it's helpful to allow them in `@objc @implementation` extensions too.)

2. **Members with an `override` modifier** override superclass members and function normally. (Again, Objective-C `@implementation` blocks often override superclass members without declaring the override in their headers, so it's helpful to allow `@objc @implementation` extensions to do the same.)

3. **Members with a `final` modifier (or `@nonobjc` on an initializer)** are Swift-only and can use Swift-only types or features. These may be Swift-only implementation details (if `internal` or `private`) or Swift-only APIs (if `public` or `package`). (This feature is necessary to support stored properties containing Swift-only types because ordinary extensions cannot declare stored properties; in theory, we could require other `final` members to be declared in a separate ordinary extension, but we also allow them in an `@objc @implementation` extension as a convenience.)
```